src/util.cpp

```diff
@@ -429,11 +429,12 @@ double Vector::Element(int i) const {
 
 bool Vector::Equals(Vector v, double tol) const {
     // Quick axis-aligned tests before going further
-    double dx = v.x - x; if(dx < -tol || dx > tol) return false;
-    double dy = v.y - y; if(dy < -tol || dy > tol) return false;
-    double dz = v.z - z; if(dz < -tol || dz > tol) return false;
+    const Vector dv = this->Minus(v);
+    if (std::abs(dv.x) > tol) return false;
```

Comment:
I don't remember if I tested this single change independently. However, you're on the right path with intrinsic: my hunch is it gives better data to the compiler and thus can be optimized better. (It also reduces operations: one comparison, and the abs which probably is a bit manipulation, vs two comparisons, but that's down to the level where I'm not sure if intuition applies when compiling with a sophisticated modern compiler)

Reply:
I have the opinon about this (ofc this should be tested): for your case dv calculated always (for all the three dimensions), which is probably resultedd in a function call (in case where inline is not performed). In case of previous implementation, the function have ability to terminate early without computation of whole vector difference.